### PR TITLE
[FIXED JENKINS-47106] Only do explicit docker pull if specified

### PR DESCRIPTION
**Pull request by abayer**
* JENKINS issue(s):
    * [JENKINS-47106](https://issues.jenkins-ci.org/browse/JENKINS-47106)
* Description:
    * Automatically doing docker pull was breaking usage of local images. It may make sense to re-enable the automatic pull for the top-level agent, though.
    * Also added a way to store arbitrary DeclarativeOptions on the DeclarativeAgent, so that we don't have to hardcode any new options in the future.
* Documentation changes:
    * https://github.com/jenkins-infra/jenkins.io/pull/1152
* Users/aliases to notify:
    * @reviewbybees

### CODE REVIEW

**Comment by abayer** [comment]
Switched to an `agent` `docker` option rather than a global `DeclarativeOption`, though I also left the logic for arbitrary `DeclarativeOption`s being pulled into `DeclarativeAgent` in place, because I can definitely see places that'd be useful.

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java [hunk 1]
```diff
@@ -66,6 +73,25 @@ public boolean isDoCheckout() {
         return doCheckout;
     }
 
+    public void addOption(@Nonnull DeclarativeOption option) {
```
Comment (rsandell) [inline pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java]: Is there any effect to this? I can't see any.

Comment (abayer) [reply pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java]: Nope, I just left it dangling. On second thought, I'll clean it out.

**Comment by abayer** [comment]
@rsandell Test updated to first build the dummy image tagged as `maven:3-alpine`, verify that without `alwaysDoPull` that's the image that gets used, and then verify that with `alwaysDoPull`, the real `maven:3-alpine` is used instead.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Agent.groovy [hunk 2]
```diff
@@ -89,6 +89,9 @@ public class Agent extends MappedClosure<Object,Agent> implements Serializable {
                 }
             }
             a.setDoCheckout(doCheckout)
+            root?.options?.options?.each { k, v ->
```
Comment (rsandell) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Agent.groovy]: Forgot to remove this?

Comment (abayer) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Agent.groovy]: argh yup

File: pipeline-model-definition/src/test/resources/dockerPullLocalImage.groovy [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            steps {
+                sh 'cat /hi-there'
+                sh 'echo "The answer is 42"'
+            }
+        }
+        stage("in pulled image") {
+            agent {
+                docker {
+                    image "maven:3-alpine"
+                    alwaysDoPull true
```
Comment (rsandell) [inline pipeline-model-definition/src/test/resources/dockerPullLocalImage.groovy]: missed renaming this

Comment (abayer) [reply pipeline-model-definition/src/test/resources/dockerPullLocalImage.groovy]: oh goddammit me.

**Pull request merged**